AMC - Enable populate template, download validation report

davidbgomes opened the pull request:
### :pushpin: References

-   **Issue:** Closes [[AMC] Add download button in the validation report](https://app.clickup.com/t/860ru2xfu)

### :memo: Implementation

- Add download button to validation report, to download submitted or calculated level excel files with populated data.

- fileType: Product Level Data
    - Download submitted data:  should download data for the current submission year for the given org unit for Program : AMC - Product Register, Stage 1 - AMC - Raw Product Consumption;
    - Download calculated data - should download data for the current submission year for the given org unit for Program : AMC - Product Register, Stage 2 - AMC - Raw Substance Consumption Calculated;

- fileType: Substance Level Data
    - Download submitted data: should download data for the current submission year for the given org unit for Program : AMC - Raw Substance Consumption Data;
    - Download calculated data - should download data for the current submission year for the given org unit for Program : AMC - Calculated Consumption Data 

### :video_camera: Screenshots/Screen capture

### :fire: Testing

9sneha-n reviewed (changes requested):
@davidbgomes , found the following issues when testing : 
1. The downloaded data has code instead of name in the excel. The same happens to download empty template as well. I dont think this is the expectation,  for example :Org Unit is "THA" but should be "Kingdom of Thailand". @ifoche  please confirm.
2. The downloaded data should be filtered by Org Unit and Period based on the selection in UI. Currently all data is being downloaded?

9sneha-n commented on src/webapp/components/current-data-submission/Validations.tsx:
```diff
@@ -1,24 +1,104 @@
+    const { currentPeriod } = useCurrentPeriodContext();
+
+    const downloadTemplate = useCallback(
+        async (downloadType: "SUBMITTED" | "CALCULATED") => {
```
This code is applicable only in AMC module, currently validation reports in other modules are not working. It continuously shows a content loader.

davidbgomes replied:
Should be fixed now

9sneha-n commented on src/domain/usecases/data-entry/egasp/GetEventProgramEmptyTemplateUseCase.ts:
```diff
@@ -1,70 +0,0 @@
-    }
-    public execute(moduleName: string, fileType: string, orgUnitId: string): FutureData<any> {
-        if (moduleName === "EGASP")
-            return this.metadataRepository.getClinicsAndLabsInOrgUnitId(orgUnitId).flatMap(orgUnits => {
```
EGASP has a requirement of showing clinics and labs in the Org Unit column of downloaded file, that whole feature is now deleted :)

davidbgomes replied:
Sorry, completely missed that. Should be fixed now. I've tested and it showed some labs and clinics in the metadata tab, so I guess it's fixed.

9sneha-n reviewed (changes requested):
1. EGASP download empty template is crashing. 
2. Validation Report for all modules other than AMC is not laoding - it shows a circular progress forever.

9sneha-n commented on src/data/dhis2/TrackedEntityInstances.ts:
```diff
@@ -0,0 +1,217 @@
+            ? { ouMode, ou: orgUnits?.map(({ id }) => id) }
+            : { ouMode };
+
+    return api.trackedEntityInstances
```
this is using old tracker api which is deprecated, glass already uses api.tracker.trackedEntities in a lot of places. please reuse that

davidbgomes replied:
Makes sense, but I'm having some issues converting this endpoint, a lot of the code is reliant on it. Maybe we can make a call and try to address this?

9sneha-n replied:
@davidbgomes Sure! create a separate PR with the new API and the error it throws ans setup a call with me.

9sneha-n commented on src/data/dhis2/TrackedEntityInstances.ts:
```diff
@@ -0,0 +1,217 @@
+    return apiTeis.map(tei => buildTei(metadata, program, tei));
+}
+
+export async function getProgram(api: D2Api, id: Id): Promise<Program | undefined> {
```
already exists in InstanceDefaultRepository: getProgram, please reuse

davidbgomes replied:
So here I have a question about this one, since this file is not a useCase, would it make sense to make use of a Repository's function? I guess the functionality is the same although written differently, but I kept it as is since I wasn't sure about this one.

9sneha-n replied:
this function is called from  DownloadTemplateDefaultRepository - we are already calling a repo from a repo. So shouldn't be an issue.

davidbgomes commented on src/webapp/components/current-data-submission/Validations.tsx:
```diff
@@ -1,24 +1,104 @@
                 </Typography>
             )}
+            <DownloadButtonsWrapper>
+                <DownloadButton
```
Makes sense, fixed it.

davidbgomes commented on src/webapp/hooks/useFileTypeByDataSubmission.ts:
```diff
@@ -0,0 +1,55 @@
+
+type GlassUploadsState = GlassState<string>;
+
+export const useFileTypeByDataSubmission = () => {
```
I've added it on the useEffect.

davidbgomes commented on src/webapp/hooks/useFileTypeByDataSubmission.ts:
```diff
@@ -0,0 +1,55 @@
+                const fileType = uploads[0]?.fileType;
+                const type =
+                    fileType === "Substance Level Data"
+                        ? "SUBSTANCE"
```
In this case, since sometimes the glass upload isn't ready in time for getting the first element and getting the file name, it could trigger an error when it was just still fetching data. That's why I fetched it as an Upload and check the loading with the ContentLoader. But you're right, the name is deceiving, I've fixed it.

davidbgomes commented on src/webapp/components/upload/UploadFiles.tsx:
```diff
@@ -327,29 +327,30 @@ export const UploadFiles: React.FC<UploadFilesProps> = ({
-            }
-        );
-    }, [compositionRoot.fileSubmission, snackbar, orgUnitId, moduleName, orgUnitCode, uploadFileType]);
+        const file = await compositionRoot.fileSubmission.downloadTemplate(api, {
```
I've tried ahah, a lot of the code on the bulk load is based on Promises, changing the logic to Futures I guess would be more trouble than it's worth. Regarding the error handling I've added try catch and tested it while solving some issues while refactoring some code, it should be good in that regard.

davidbgomes commented on src/data/repositories/download-template/Dhis2RelationshipTypes.ts:
```diff
@@ -0,0 +1,334 @@
+
+export type RelationshipOrgUnitFilter = TrackedEntityOURequestApi["ouMode"];
+
+export function getApiRelationships(
```
Nice catch, yeah I've copied the file but those two functions never got used. Done

davidbgomes commented:
Hello @9sneha-n , thanks a lot for the thorough review! I've fixed the major issues you pointed, although I've left some comments on some of the issues you pointed. Let me know what you think on those, and maybe we can make a call to address it. Thanks again

9sneha-n reviewed (changes requested):
@davidbgomes the following use cases still dont work as expected : 
1. AMC Product level data - Download empty template : Downloaded file populates data as code, whereas name is expected. This was the earlier behaviour.
2. AMC Substance level data - Download empty template : This is a Event Program, but the downloaded file has 2 tabs like in tracker program. This is the wrong format. 
Additionally, the file populates fields as codes instead of names.
3. EGASP - Download empty template : : This is a Event Program, but the downloaded file has 2 tabs like in tracker program. This is the wrong format. 
4. Download Submitted Data 
a. For a data submission with correct product level upload (DZA, 2018) - Error downloading file
b. For a data submission with correct substance level upload (DZA, 2017) - Error downloading file
c. If no  correct data uploaded for a data submission - the downloaded file is incorrect. It should be empty as no correct upload exists.

5. Download Calculated Data 
a. For a data submission with correct product level upload (DZA, 2018) - Error downloading file
b. For a data submission with correct substance level upload (DZA, 2017) - Error downloading file
c. If no  correct data uploaded for a data submission - the downloaded file is incorrect. It should be empty as no correct upload exists.


6. Whenever a file is downloaded for AMC validation report, the file format is incorrect. The second tab should be AMC-product register stage 1 for Submitted data and AMC- product register - Stage 2 for calculated data if uploaded data is product level. 

8. all event programs are downloaded as tracker programs which is incorrect (DZA , 2022)

davidbgomes commented:
Thanks @9sneha-n , I've addressed the issues on this PR and the ones we talked about:
- EGASP using metadata codes instead of names;
- AMC populate, filter TEI's by start and end date as well

9sneha-n reviewed (changes requested):
@davidbgomes request the following changes
1. AMC Download Calculated data for Product level is incorrect. The second tab should show product calculated data but is showing empty raw substance data.
2. Download empty template AMC Product level has the same issue. The third tab is showing has AMC Raw Substance details.
3. For thailand 2020, although i can see an completed upload, the download submitted data and download calculated dat buttons are disabled.
4. The downloaded file names say template instead of "submitted-data" or "calculated-data"

9sneha-n commented on package.json:
```diff
@@ -32,6 +32,7 @@
         "font-awesome": "4.7.0",
         "lodash": "4.17.21",
         "lodash.product": "^18.10.0",
+        "moment": "^2.30.1",
```
isnt moment already available? its used in the code already, this might not be neccessary

davidbgomes replied:
Yes.. I just checked and it is available, I don't know how though. Anyways I think it's better to have it in package.json either way, so it's clear for someone new that we use that package. What do you think?

9sneha-n commented on src/domain/usecases/data-entry/DownloadTemplateUseCase.ts:
```diff
@@ -0,0 +1,197 @@
+
+        const element = await this.DownloadtemplateRepository.getElement(formType, programId);
+
+        let orgUnits = [orgUnit];
```
do not use let.  
const orgUnits = moduleName === "EGASP" ? clinicsAndLabsOrgUnits : [orgUnit]

9sneha-n commented:
> @davidbgomes request the following changes
> 
> 1. AMC Download Calculated data for Product level is incorrect. The second tab should show product calculated data but is showing empty raw substance data.
> 2. Download empty template AMC Product level has the same issue. The third tab is showing has AMC Raw Substance details.
> 3. For thailand 2020, although i can see an completed upload, the download submitted data and download calculated dat buttons are disabled.
> 4. The downloaded file names say template instead of "submitted-data" or "calculated-data"

@davidbgomes Apologies! Please ignore points #1 and #2. I got confused between the names, super sorry! The template and name of the calculated data in the downloaded file is correct. However: 
1. The calculated data is not getting populated for Product level.  Even though i see the calculation in DHIS, the downloaded file has empty 2nd tab.(Thailand, 2017,2023 )
2. A strange error - In Thailand , 2023 -> Download Submitted data, the data is not filtered by OU or Period.
